pkg/blockbuilder/scheduler/queue.go

```diff
@@ -124,237 +120,216 @@ func (q *JobQueue) RunLeaseExpiryChecker(ctx context.Context) {
-	case types.JobStatusComplete:
-		return q.completed.Lookup(func(jwm *JobWithMetadata) bool {
-			return jwm.ID() == id
-		})
-	default:
-		return nil, false
-	}
+// jobIDExtractor extracts the job ID from a JobWithMetadata
+func jobIDExtractor(j *JobWithMetadata) string {
+	return j.ID()
 }
 
-// Enqueue adds a job to the pending queue with the given priority
-func (q *JobQueue) Enqueue(job *types.Job, priority int) error {
+// TransitionState attempts to transition a job from one specific state to another
+func (q *JobQueue) TransitionState(jobID string, from, to types.JobStatus) (bool, error) {
 	q.mu.Lock()
 	defer q.mu.Unlock()
 
-	return q.enqueueLockLess(job, priority)
-}
+	currentStatus, exists := q.statusMap[jobID]
+	if !exists {
+		return false, fmt.Errorf("job %s not found", jobID)
+	}
 
-func (q *JobQueue) enqueueLockLess(job *types.Job, priority int) error {
-	// Check if job already exists
-	if status, exists := q.statusMap[job.ID()]; exists && status != types.JobStatusExpired {
-		return fmt.Errorf("job %s already exists with status %v", job.ID(), status)
+	if currentStatus != from {
+		return false, fmt.Errorf("job %s is in state %s, not %s", jobID, currentStatus, from)
 	}
 
-	jobMeta := NewJobWithMetadata(job, priority)
-	q.pending.Push(jobMeta)
-	q.statusMap[job.ID()] = types.JobStatusPending
-	q.metrics.pending.Inc()
-	return nil
+	return q.transitionLockLess(jobID, to)
 }
 
-// Dequeue removes and returns the highest priority job from the pending queue
-func (q *JobQueue) Dequeue() (*types.Job, bool) {
+// TransitionAny transitions a job from any state to the specified state
+func (q *JobQueue) TransitionAny(jobID string, to types.JobStatus, createFn func() (*JobWithMetadata, error)) (prevStatus types.JobStatus, found bool, err error) {
 	q.mu.Lock()
 	defer q.mu.Unlock()
 
-	jobMeta, ok := q.pending.Pop()
-	if !ok {
-		return nil, false
-	}
-	q.metrics.pending.Dec()
+	currentStatus, exists := q.statusMap[jobID]
+	if finished := currentStatus.IsFinished(); !exists || finished {
+
+		// we're just moving one finished type to another; no need to re-enqueue
+		if finished && to.IsFinished() {
+			q.statusMap[jobID] = to
+			if j, found := q.completed.Lookup(
+				func(jwm *JobWithMetadata) bool {
+					return jwm.ID() == jobID
+				},
+			); found {
+				j.Status = to
+				j.UpdateTime = time.Now()
+			}
+			return currentStatus, true, nil
+		}
 
-	// Update metadata for in-progress state
-	jobMeta.Status = types.JobStatusInProgress
-	jobMeta.StartTime = time.Now()
-	jobMeta.UpdateTime = jobMeta.StartTime
+		if createFn == nil {
+			return types.JobStatusUnknown, false, fmt.Errorf("job %s not found and no creation function provided", jobID)
+		}
 
-	q.inProgress[jobMeta.ID()] = jobMeta
-	q.statusMap[jobMeta.ID()] = types.JobStatusInProgress
-	q.metrics.inProgress.Inc()
+		if finished {
+			level.Debug(q.logger).Log("msg", "ignoring transition for completed job; will recreate", "id", jobID, "from", currentStatus, "to", to)
```

Comment:
is this comment correct?

Reply:
Yes, but perhaps I can be clearer: we're not going to delete the old completed entry for this job. Instead we're going to recreate a copy.